### PR TITLE
Breeze & Runner, Runner Mini support

### PR DESCRIPTION
**Pull request by jafar-atili**
# Pull Request

> First PR? don't hesitate, push it.

## Description

> Describe what you did and why.

- Added support for Switcher Breeze device
- Added support for Switcher Runner devices

## Checklist

- [x] I have followed this repository's contributing guidelines.
- [x] I have used the conventional commits specification for my commit messages.
- [x] I have signed my commits.
- [x] I will adhere to the project's code of conduct.

## Additional information

> Anything else?
>

### CODE REVIEW

File: src/aioswitcher/api/messages.py [hunk 1]
```diff
@@ -78,6 +85,63 @@ def get_state(self) -> DeviceState:
...
+        hex_level = self._hex_response[162:163].decode()
+        levels = dict(map(lambda s: (s.value, s), ThermostatFanLevel))
+        try:
+            return levels[hex_level]
+        except KeyError:
+            return ThermostatFanLevel.LOW
+
+    def get_thermostat_swing(self) -> ThermostatSwing:
+        """Return the current thermostat fan swing."""
+        hex_swing = self._hex_response[162:161].decode()
```
Comment (thecode) [inline src/aioswitcher/api/messages.py]: Fix swing:
```suggestion
        hex_swing = self._hex_response[163:164].decode()
```

File: src/aioswitcher/device/__init__.py [hunk 2]
```diff
@@ -181,3 +323,26 @@ def __post_init__(self) -> None:
...
+
+
+@final
+@dataclass
+class SwitcherShutter(SwitcherShutterBase, SwitcherBase):
+    """Implementation of the Switcher Shutter device."""
+
+    def __post_init__(self) -> None:
+        if self.device_type.category != DeviceCategory.SHUTTER:
+            raise ValueError("only shutter are allowed")
```
Comment (thecode) [inline src/aioswitcher/device/__init__.py]:
```suggestion
            raise ValueError("only shutters are allowed")
```

File: src/aioswitcher/device/__init__.py [hunk 3]
```diff
@@ -181,3 +323,26 @@ def __post_init__(self) -> None:
...
+
+
+@final
+@dataclass
+class SwitcherThermostat(SwitcherThermostatBase, SwitcherBase):
+    """Implementation of the Switcher Thermostat device."""
+
+    def __post_init__(self) -> None:
+        if self.device_type.category != DeviceCategory.THERMOSTAT:
+            raise ValueError("only thermostat are allowed")
```
Comment (thecode) [inline src/aioswitcher/device/__init__.py]:
```suggestion
            raise ValueError("only thermostats are allowed")
```

File: src/aioswitcher/device/__init__.py [hunk 4]
```diff
@@ -91,6 +96,109 @@ def value(self) -> str:
...
+    def value(self) -> str:
+        """Return the value of the fan level."""
+        return self._value  # type: ignore
+
+
+class ThermostatSwing(Enum):
+    """Enum class representing the thermostat device's swing state."""
+
+    OFF = "0", "off"
+    ON = "1", "on"  # TODO: Get actual Value
```
Comment (thecode) [inline src/aioswitcher/device/__init__.py]:
```suggestion
    ON = "1", "on"
```
Verified it is working

**Review by thecode — COMMENTED** [review]
First quick pass, looks great.

- Please remove: `src/aioswitcher/api/.DS_Store`
- Undo changes on `scripts/discover_devices.py]` (for some reason git shows as an empty file

**Comment by jafar-atili** [comment]
> First quick pass, looks great.
> 
> * Please remove: `src/aioswitcher/api/.DS_Store`
> * Undo changes on `scripts/discover_devices.py]` (for some reason git shows as an empty file

Done

File: src/aioswitcher/api/__init__.py [hunk 5]
```diff
@@ -325,3 +450,320 @@ async def create_schedule(
...
+        form = FormData()
+        form.add_field("token", "d41d8cd98f00b204e9800998ecf8427e")
+        form.add_field("rtps", hexlify(udp_message.unparsed_response).decode())
+        async with ClientSession() as session:
+            async with session.post(
+                "https://switcher.co.il/misc/irGet/getIR.php", data=form
+            ) as resp:
+                if resp.status == 200:
+                    logger.debug(
+                        "IR Set file succesfully downloaded from Switcher (length=%s)",
```
Comment (thecode) [inline src/aioswitcher/api/__init__.py]:
```suggestion
                        "IR Set file successfully downloaded from Switcher (length=%s)",
```

File: src/aioswitcher/api/packets.py [hunk 6]
```diff
@@ -17,16 +17,29 @@
 # weekdays sum, start-time timestamp, end-time timestamp
 SCHEDULE_CREATE_DATA_FORMAT = "01{}01{}{}"
 
+PHONE_ID = "0000"
+DEVICE_PASSWORD = "00000000"
```
Comment (thecode) [inline src/aioswitcher/api/packets.py]: bandit is complaining about hardcoded password and we don't use the `PHONE_ID` and `DEVICE_PASSWORD` anyhow so we can do the same as we do for other devices by replacing all zeros with padding:
We can update also the older ones to use 72 padded zeros and all packets will have the same paddings:

`SEND_CONTROL_PACKET = "fef05d0002320102" + REQUEST_FORMAT + "{}" + PAD_72_ZEROS + "000106000{}00{}"`

File: src/aioswitcher/api/packets.py [hunk 7]
```diff
@@ -17,16 +17,29 @@
...
 NO_TIMER_REQUESTED = "00000000"
 NON_RECURRING_SCHEDULE = "00"
 # format values are local session id, timestamp
 REQUEST_FORMAT = "{}340001000000000000000000{}00000000000000000000f0fe"
-PAD_74_ZEROS = "0" * 74
+REQUEST_FORMAT2 = "{}390001000000000000000000{}00000000000000000000f0fe"
 
+REQUEST_FORMAT_BREEZE = "{}000001000000000000000000{}00000000000000000000f0fe"
+PAD_74_ZEROS = "0" * 74
```
Comment (thecode) [inline src/aioswitcher/api/packets.py]:
```suggestion
PAD_72_ZEROS = "0" * 72
```

File: src/aioswitcher/api/packets.py [hunk 8]
```diff
@@ -17,16 +17,29 @@
...
+
 # format values are local session id, timestamp, device id
 GET_STATE_PACKET = "fef0300002320103" + REQUEST_FORMAT + "{}00"
+
+# format values are local session id, timestamp, device id
+GET_STATE_PACKET2 = "fef0300003050103" + REQUEST_FORMAT2 + "{}00"
+
 # format values are local session id, timestamp, device id, command, timer
 SEND_CONTROL_PACKET = (
     "fef05d0002320102" + REQUEST_FORMAT + "{}" + PAD_74_ZEROS + "0106000{}00{}"
```
Comment (thecode) [inline src/aioswitcher/api/packets.py]:
```suggestion
    "fef05d0002320102" + REQUEST_FORMAT + "{}" + PAD_72_ZEROS + "000106000{}00{}"
```
We need to do the same fix for all the others bellow (can't suggest them on GitHub)

File: src/aioswitcher/api/packets.py (hunk 6 above)
Comment (thecode) [inline src/aioswitcher/api/packets.py]:
```suggestion
```

File: src/aioswitcher/api/packets.py [hunk 9]
```diff
@@ -53,3 +66,48 @@
...
+    "fef0000003050102"
+    + REQUEST_FORMAT_BREEZE
+    + "{}"
+    + "00"
+    + PHONE_ID
+    + "0000"
+    + DEVICE_PASSWORD
+    + "0000000000000000000000000000000000000000000000000000003701"
+    + "{}{}"
+)
```
Comment (thecode) [inline src/aioswitcher/api/packets.py]:
```suggestion
BREEZE_COMMAND_PACKET = (
    "fef0000003050102" + REQUEST_FORMAT_BREEZE + "{}" + PAD_72_ZEROS + +"3701" + "{}{}"
)
```

File: src/aioswitcher/api/packets.py [hunk 10]
```diff
@@ -53,3 +66,48 @@
...
+    + "{}"
+    + "00000000000000000000f0fe"
+    + "{}"
+    + "00"
+    + PHONE_ID
+    + "0000"
+    + DEVICE_PASSWORD
+    + "0000000000000000000000000000000000000000000000000000003702"
+    + "02000000"
+)
```
Comment (thecode) [inline src/aioswitcher/api/packets.py]:
```suggestion
RUNNER_STOP_COMMAND = (
    "fef0590003050102" + "{}" + "232301" 
    + "000000000000000000"
    + "{}"
    + "00000000000000000000f0fe"
    + "{}"
    + PAD_72_ZEROS +
	+ "370202000000"
)
```

File: src/aioswitcher/api/packets.py [hunk 11]
```diff
@@ -53,3 +66,48 @@
...
+    + "00000000000000000000f0fe"
+    + "{}"
+    + "00"
+    + PHONE_ID
+    + "0000"
+    + DEVICE_PASSWORD
+    + "0000000000000000000000000000000000000000000000000000003701"
+    + "0100"
+    + "{}"
+)
```
Comment (thecode) [inline src/aioswitcher/api/packets.py]:
```suggestion
RUNNER_SET_POSITION = (
    "fef0580003050102{}290401"
    + "000000000000000000"
    + "{}"
    + "00000000000000000000f0fe"
    + "{}"
    + PAD_72_ZEROS +
	+ "3701"
    + "0100"
    + "{}"
)
```

**Review by thecode — COMMENTED** [review]
I added some suggestions how to remove the use of `PHONE_ID` and `DEVICE_PASSWORD` , it will also fix the bandit failure

File: src/aioswitcher/api/__init__.py (hunk 5 above)
Comment (thecode) [inline src/aioswitcher/api/__init__.py]:
```suggestion
                        "IR Set file successfully downloaded from Switcher (length=%s)",
```

File: src/aioswitcher/api/packets.py [hunk 12]
```diff
@@ -21,35 +21,74 @@
...
+
+RUNNER_SET_POSITION = (
+    "fef0580003050102{}290401"
+    + "000000000000000000"
+    + "{}"
+    + "00000000000000000000f0fe"
+    + "{}"
+    + PAD_72_ZEROS +
+	+ "3701"
```
Comment (thecode) [inline src/aioswitcher/api/packets.py]:
```suggestion
    "fef0590003050102"
    + "{}"
    + "232301"
    + "000000000000000000"
    + "{}"
    + "00000000000000000000f0fe"
    + "{}"
    + PAD_72_ZEROS
    + "370202000000"
)

RUNNER_SET_POSITION = (
    "fef0580003050102{}290401"
    + "000000000000000000"
    + "{}"
    + "00000000000000000000f0fe"
    + "{}"
    + PAD_72_ZEROS
    + "3701"
```

File: src/aioswitcher/api/packets.py [hunk 13]
```diff
@@ -21,35 +21,76 @@
 NON_RECURRING_SCHEDULE = "00"
 # format values are local session id, timestamp
 REQUEST_FORMAT = "{}340001000000000000000000{}00000000000000000000f0fe"
-PAD_74_ZEROS = "0" * 74
+REQUEST_FORMAT2 = "{}390001000000000000000000{}00000000000000000000f0fe"
 
+REQUEST_FORMAT_BREEZE = "{}000001000000000000000000{}00000000000000000000f0fe"
+PAD_72_ZEROS = "0" * 72
 # format value just timestamp (initial session id is "00000000")
 LOGIN_PACKET = "fef052000232a10000000000" + REQUEST_FORMAT[2:] + "1c" + PAD_74_ZEROS
```
Comment (thecode) [inline src/aioswitcher/api/packets.py]:
```suggestion
LOGIN_PACKET = (
    "fef052000232a10000000000" + REQUEST_FORMAT[2:] + "1c" + PAD_72_ZEROS + "00"
)
```

File: src/aioswitcher/api/packets.py [hunk 14]
```diff
@@ -21,35 +21,78 @@
...
 # schedule data =
 #                   (on_off + week + timstate + start_time + end_time)
 CREATE_SCHEDULE_PACKET = (
-    "fef0630002320102" + REQUEST_FORMAT + "{}" + PAD_74_ZEROS + "030c00ff{}"
+    "fef0630002320102" + REQUEST_FORMAT + "{}" + PAD_72_ZEROS + "00030c00ff{}"
+)
+# format values are local session id, timestamp, device id, phone id, device-
+# passwored, command length, command
+BREEZE_COMMAND_PACKET = (
+    "fef0000003050102" + REQUEST_FORMAT_BREEZE + "{}" + PAD_72_ZEROS + +"3701" + "{}{}"
```
Comment (thecode) [inline src/aioswitcher/api/packets.py]:
```suggestion
    "fef0000003050102" + REQUEST_FORMAT_BREEZE + "{}" + PAD_72_ZEROS + "3701" + "{}{}"
```

File: .vscode/settings.json [hunk 15]
```diff
@@ -1,3 +1,5 @@
 {
-    "python.pythonPath": "${workspaceFolder}/.venv/bin/python"
+    "python.pythonPath": "${workspaceFolder}/.venv/bin/python",
+    "python.linting.flake8Enabled": true,
```
Comment (TomerFi) [inline .vscode/settings.json]: I think these better fit the user's settings.

File: src/aioswitcher/device/__init__.py [hunk 16]
```diff
@@ -151,6 +259,40 @@ def auto_off_set(self) -> str:
...
+        target temprature: the current target temprature in celsius.
+        fan_level: the current fan level in celsius.
+        swing: the current swing state.
+        remote_id: the id of the remote used to control this thermostat
+    """
+
+    mode: ThermostatMode
+    temprature: float
+    target_temprature: int
+    fan_leve: ThermostatFanLevel
```
Comment (thecode) [inline src/aioswitcher/device/__init__.py]:
```suggestion
    fan_level: ThermostatFanLevel
```

File: src/aioswitcher/api/packets.py [hunk 17]
```diff
@@ -21,35 +21,78 @@
 NON_RECURRING_SCHEDULE = "00"
 # format values are local session id, timestamp
 REQUEST_FORMAT = "{}340001000000000000000000{}00000000000000000000f0fe"
-PAD_74_ZEROS = "0" * 74
+REQUEST_FORMAT2 = "{}390001000000000000000000{}00000000000000000000f0fe"
```
Comment (TomerFi) [inline src/aioswitcher/api/packets.py:24]: we need more meaningful variable names here.

File: src/aioswitcher/api/packets.py [hunk 18]
```diff
@@ -21,35 +21,78 @@
...
 
+REQUEST_FORMAT_BREEZE = "{}000001000000000000000000{}00000000000000000000f0fe"
+PAD_72_ZEROS = "0" * 72
 # format value just timestamp (initial session id is "00000000")
-LOGIN_PACKET = "fef052000232a10000000000" + REQUEST_FORMAT[2:] + "1c" + PAD_74_ZEROS
+LOGIN_PACKET = (
+    "fef052000232a10000000000" + REQUEST_FORMAT[2:] + "1c" + PAD_72_ZEROS + "00"
+)
+
+LOGIN2_PACKET = (
```
Comment (TomerFi) [inline src/aioswitcher/api/packets.py:33]: we need more meaningful variable names here.

File: src/aioswitcher/api/packets.py [hunk 19]
```diff
@@ -21,35 +21,78 @@
...
+    "fef030000305a60000000000ff0301000000000000000000{}00000000000000000000f0fe{}00"
+)
+
 # format values are local session id, timestamp, device id
 GET_STATE_PACKET = "fef0300002320103" + REQUEST_FORMAT + "{}00"
+
+# format values are local session id, timestamp, device id
+GET_STATE_PACKET2 = "fef0300003050103" + REQUEST_FORMAT2 + "{}00"
```
Comment (TomerFi) [inline src/aioswitcher/api/packets.py:41]: we need more meaningful variable names here.

File: src/aioswitcher/bridge.py [hunk 20]
```diff
@@ -29,16 +29,33 @@
...
+    ThermostatSwing,
 )
 from .device.tools import seconds_to_iso_time, watts_to_amps
 
 __all__ = ["SwitcherBridge"]
 logger = getLogger(__name__)
 
 
+SWITCHER_UDP_PORT = 20002
+SWITCHER_UDP_PORT2 = 20003
```
Comment (TomerFi) [inline src/aioswitcher/bridge.py:49]: we need more meaningful variable names here.

File: src/aioswitcher/bridge.py [hunk 21]
```diff
@@ -218,6 +274,12 @@ def get_ip(self) -> str:
         ip_addr = int(hex_ip[6:8] + hex_ip[4:6] + hex_ip[2:4] + hex_ip[0:2], 16)
         return inet_ntoa(pack("<L", ip_addr))
 
+    def get_ip2(self) -> str:
```
Comment (TomerFi) [inline src/aioswitcher/bridge.py:277]: we need more meaningful function names here.

**Review by TomerFi — CHANGES_REQUESTED** [review]
@jafar-atili thank you very much, this looks very good and quickly shows you put a lot of effort into it.

I added some small change requests, nothing too big.
but, as this is a pretty big pr I wanna review it some more,
and perhaps I'll have a couple of other crs/questions.

**Comment by TomerFi** [comment]
@jafar-atili 

Some of the tests now fail based on these modifications, we need to fix them,
and we also need to add some tests for the new code.

In regards to commit messages, please amend them based on the [conventional commits specification](https://www.conventionalcommits.org/en/v1.0.0/), and also sign them, using _git_'s _signoff_ option or manually to indicate you acknowledge [developer certificate of origin document](https://developercertificate.org/).

File: src/aioswitcher/api/__init__.py [hunk 22]
```diff
@@ -325,3 +469,264 @@ async def create_schedule(
...
+
+            temp = key[2:4]
+            if temp.isdigit():
+                temp = int(temp)
+                if temp > self.cap_max_temp:
+                    self.cap_max_temp = temp
+                if temp < self.cap_min_temp:
+                    self.cap_min_temp = temp
+
+            self.cap_swingable = not self.cap_swingable and "d1" in key
```
Comment (thecode) [inline src/aioswitcher/api/__init__.py]:
```suggestion
            self.cap_swingable |= "d1" in key
```
Using not will set it back to false on next iteration

File: .vscode/settings.json (hunk 15 above)
Comment (thecode) [inline .vscode/settings.json]:
```suggestion
```

File: .vscode/settings.json [hunk 23]
```diff
@@ -1,3 +1,3 @@
 {
-    "python.pythonPath": "${workspaceFolder}/.venv/bin/python"
+    "python.linting.enabled": true
```
Comment (thecode) [inline .vscode/settings.json]:
```suggestion
    "python.pythonPath": "${workspaceFolder}/.venv/bin/python"
```

**Comment by thecode** [comment]
Too much for editing from the git cli, the following files as a typo: `temprature` ==> `temperature`:
`control_device.py`, `messages.py`, `device/__init__.py`

File: src/aioswitcher/api/__init__.py (hunk 22 above)
Comment (jafar-atili) [reply src/aioswitcher/api/__init__.py]: Correct, However in my IRSet it looks different.

There are two keys FUN_d1 and FUN_d0

In Switcher app if I press the swing, it'll swing but it shows in the APP that the swing is Moving but after few seconds it shows Stop.

I believe for such AC the swing command should be handled differently and in a specific command.

Comment (jafar-atili) [reply src/aioswitcher/api/__init__.py]: I guess I most go back with this to them and ask how it can be handled in single command.

**Review by thecode — APPROVED** [review]
Current code coverage is 75.17%, follow up PRs will add tests coverage

**Comment by TomerFi** [comment]
@all-contributors please add @jafar-atili for code and doc

**Pull request merged**